desc/integrals/bounce_utils.py

```diff
@@ -0,0 +1,809 @@
+    )
+    assert intersect.shape == (P, S, knots.size - 1, degree)
+
+    # Reshape so that last axis enumerates intersects of a pitch along a field line.
+    dB_dz_sign = jnp.sign(
+        polyval_vec(x=intersect, c=dB_dz[..., jnp.newaxis]).reshape(P, S, -1)
+    )
+    # Only consider intersect if it is within knots that bound that polynomial.
+    is_intersect = intersect.reshape(P, S, -1) >= 0
+    # Following discussion on page 3 and 5 of https://doi.org/10.1063/1.873749,
+    # we ignore the bounce points of particles only assigned to a class that are
+    # trapped outside this snapshot of the field line.
+    is_z1 = (dB_dz_sign <= 0) & is_intersect
+    is_z2 = (dB_dz_sign >= 0) & _in_epigraph_and(is_intersect, dB_dz_sign)
+
+    # Transform out of local power basis expansion.
+    intersect = (intersect + knots[:-1, jnp.newaxis]).reshape(P, S, -1)
+    # New versions of JAX only like static sentinels.
+    sentinel = -10000000.0  # instead of knots[0] - 1
+    z1 = take_mask(intersect, is_z1, size=num_well, fill_value=sentinel)
+    z2 = take_mask(intersect, is_z2, size=num_well, fill_value=sentinel)
+
+    mask = (z1 > sentinel) & (z2 > sentinel)
+    # Set outside mask to same value so integration is over set of measure zero.
+    z1 = jnp.where(mask, z1, 0.0)
+    z2 = jnp.where(mask, z2, 0.0)
+
+    if check:
+        _check_bounce_points(z1, z2, pitch, knots, B, plot, **kwargs)
+
+    return z1, z2
+
+
+def _check_bounce_points(z1, z2, pitch, knots, B, plot=True, **kwargs):
+    """Check that bounce points are computed correctly."""
+    eps = kwargs.pop("eps", jnp.finfo(jnp.array(1.0).dtype).eps * 10)
+    kwargs.setdefault(
+        "title",
+        r"Intersects $\zeta$ in epigraph($\vert B \vert$) s.t. "
+        r"$\vert B \vert(\zeta) = 1/\lambda$",
+    )
+    kwargs.setdefault("klabel", r"$1/\lambda$")
+    kwargs.setdefault("hlabel", r"$\zeta$")
+    kwargs.setdefault("vlabel", r"$\vert B \vert$")
+
+    assert z1.shape == z2.shape
+    mask = (z1 - z2) != 0.0
+    z1 = jnp.where(mask, z1, jnp.nan)
+    z2 = jnp.where(mask, z2, jnp.nan)
+
+    err_1 = jnp.any(z1 > z2, axis=-1)
+    err_2 = jnp.any(z1[..., 1:] < z2[..., :-1], axis=-1)
+
+    P, S, _ = z1.shape
+    for s in range(S):
+        Bs = PPoly(B[:, s], knots)
+        for p in range(P):
+            Bs_midpoint = Bs((z1[p, s] + z2[p, s]) / 2)
+            err_3 = jnp.any(Bs_midpoint > 1 / pitch[p, s] + eps)
+            if not (err_1[p, s] or err_2[p, s] or err_3):
+                continue
+            _z1 = z1[p, s][mask[p, s]]
+            _z2 = z2[p, s][mask[p, s]]
+            if plot:
+                plot_ppoly(
+                    ppoly=Bs,
+                    z1=_z1,
+                    z2=_z2,
+                    k=1 / pitch[p, s],
+                    **kwargs,
+                )
+
+            print("      z1    |    z2")
+            print(jnp.column_stack([_z1, _z2]))
+            assert not err_1[p, s], "Intersects have an inversion.\n"
```

Comment:
I recommend this as it is. The plots are shown only when there is an error and before the assert statements trigger.

> count up the number of errors then print them all at the end
It would be challenging to understand what went wrong then